utils/compilation_filter.py

```diff
@@ -0,0 +1,265 @@
+
+    Public attributes:
+        addedNodes_: List[str]. Added(created) nodes of this transformation.
+        removedNodes_: List[str]. Removed(replaced) nodes of this transformation.
+        ancestors_: List['Transformation']. The ancestor transformation of current transformation.
+        scopeName_: str. The scope of current transformation.
+        transID_: str. The internal transformation ID in the database.
+        isDirectTrans_ :bool. Whether this transformation directly created/replaced the given nodeName that is passed to this script file.
+    """
+
+    def __init__(self, transID: str):
+        self.addedNodes_: List[str] = []
+        self.removedNodes_: List[str] = []
+        self.ancestors_: List['Transformation'] = []
+        self.scopeName_: str = ''
+        self.transID_: str = transID
+        self.isDirectTrans_: bool = False
+
+    def appendAddedNode(self, nodeName: str) -> None:
+        """Append the added nodes of this transformation."""
+
+        self.addedNodes_.append(nodeName)
+
+    def appendRemovedNode(self, nodeName: str) -> None:
+        """Append the removed nodes of this transformation."""
+
+        self.removedNodes_.append(nodeName)
+
+    def addAnc(self, anc: 'Transformation') -> None:
+        """Add ancestors of this transformation."""
+
+        self.ancestors_.append(anc)
+
+
+class DottyPrinter:
+    """A class for generating the dotty graph file"""
+
+    def __init__(self):
+        self.vertices_ = []
+        self.edges_ = []
+
+    def get_color(self, isDirectTrans: bool) -> str:
+        """Returns the color for the given node. """
+
+        if isDirectTrans:
+            return "Yellow2"
+        else:
+            return "AliceBlue"
+
+    def dump_label(self, tran: Transformation) -> str:
+        """Returns the string for the label of the given transformation. """
+
+        labelStr = rf"""{{ {{SCOPE:\l{tran.scopeName_} }}|{{ORIGINAL NODES:\l\l"""
+        for rstr in tran.removedNodes_:
+            labelStr += rf"""{rstr}\l\l"""
+        labelStr += rf"}}| {{REPLACED BY:\l\l"
+        for astr in tran.addedNodes_:
+            labelStr += rf"""{astr}\l\l"""
+        labelStr += f"}} }}"
+        return labelStr
+
+    def dump_node(self, tran: Transformation) -> None:
+        """Generates the dotty information for the given transformation. """
+
+        if not tran:
+            return
+
+        tranStr = f"""v{tran.transID_}[\n
+                    \tlabel = \"{self.dump_label(tran)}\"\n
+                    \tshape = \"record\"\n
+                    \tstyle=\"filled,rounded\"\n
+                    \tfillcolor={self.get_color(tran.isDirectTrans_)}\n
+                    penwidth = 2];\n"""
+        self.vertices_.append(tranStr)
+
+    def visit_nodes(self) -> None:
+        """Visits all transformation and dump the dotty information for each transformation. """
+
+        for tran in TRANS_LIST:
```

Comment:
I guess I should have said this on the last PR but is there a way to avoid passing data to the dot graph printer class through global variables?